arch/xtensa/src/common/xtensa_blocktask.c

```diff
@@ -45,108 +45,69 @@
+  nxsched_suspend_scheduler(rtcb);
 
-  DEBUGASSERT((tcb->task_state >= FIRST_READY_TO_RUN_STATE) &&
-              (tcb->task_state <= LAST_READY_TO_RUN_STATE));
+  /* Are we in an interrupt handler? */
 
-  /* Remove the tcb task from the ready-to-run list.  If we are blocking the
-   * task at the head of the task list (the most likely case), then a
-   * context switch to the next ready-to-run task is needed. In this case,
-   * it should also be true that rtcb == tcb.
-   */
-
-  switch_needed = nxsched_remove_readytorun(tcb);
-
-  /* Add the task to the specified blocked task list */
-
-  nxsched_add_blocked(tcb, (tstate_t)task_state);
-
-  /* If there are any pending tasks, then add them to the ready-to-run
-   * task list now
-   */
-
-  if (g_pendingtasks.head)
-    {
-      switch_needed |= nxsched_merge_pending();
-    }
-
-  /* Now, perform the context switch if one is needed */
-
-  if (switch_needed)
+  if (CURRENT_REGS)
     {
-      /* Update scheduler parameters */
-
-      nxsched_suspend_scheduler(rtcb);
-
-      /* Are we in an interrupt handler? */
+      /* Yes, then we have to do things differently.
+       * Just copy the CURRENT_REGS into the OLD rtcb.
+       */
 
-      if (CURRENT_REGS)
-        {
-          /* Yes, then we have to do things differently.
-           * Just copy the CURRENT_REGS into the OLD rtcb.
-           */
+      xtensa_savestate(rtcb->xcp.regs);
 
-          xtensa_savestate(rtcb->xcp.regs);
+      /* Restore the exception context of the rtcb at the (new) head
+       * of the ready-to-run task list.
+       */
 
-          /* Restore the exception context of the rtcb at the (new) head
-           * of the ready-to-run task list.
-           */
+      rtcb = this_task();
 
-          rtcb = this_task();
+      /* Reset scheduler parameters */
 
-          /* Reset scheduler parameters */
+      nxsched_resume_scheduler(rtcb);
 
-          nxsched_resume_scheduler(rtcb);
+      /* Then switch contexts.  Any necessary address environment
+       * changes will be made when the interrupt returns.
+       */
 
-          /* Then switch contexts.  Any necessary address environment
-           * changes will be made when the interrupt returns.
-           */
-
-          xtensa_restorestate(rtcb->xcp.regs);
-        }
+      xtensa_restorestate(rtcb->xcp.regs);
+    }
 
-      /* No, then we will need to perform the user context switch */
+  /* No, then we will need to perform the user context switch */
 
-      else
-        {
-          struct tcb_s *nexttcb = this_task();
+  else
+    {
+      struct tcb_s *nexttcb = this_task();
 
-          /* Switch context to the context of the task at the head of the
-           * ready to run list.
-           */
+      /* Switch context to the context of the task at the head of the
+       * ready to run list.
+       */
 
-          nxsched_resume_scheduler(nexttcb);
+      nxsched_resume_scheduler(nexttcb);
```

Comment:
Do we need
```
#ifdef CONFIG_ARCH_ADDRENV
              /* Make sure that the address environment for the previously
               * running task is closed down gracefully (data caches dump,
               * MMU flushed) and set up the address environment for the new
               * thread at the head of the ready-to-run list.
               */

              group_addrenv(nexttcb);
#endif
```
before calling `nxsched_resume_scheduler`?

Reply:
I guess xtensa doesn't support kernel mode yet. And the origin code doens't invoke group_addrenv, it's better to not add group_addrenv in this patch.